### PR TITLE
Introduce AdvectionField::scalar_extractor (fix issue 755)

### PR DESCRIPTION
**Pull request by yinghe616**
This patch is aimed at addressing issue #755 . It's ready for review.

### CODE REVIEW

File: include/aspect/simulator.h [hunk 1]
```diff
@@ -263,6 +263,14 @@ namespace aspect
          * See Introspection::base_elements for more information.
          */
         unsigned int base_element(const Introspection<dim> &introspection) const;
+
+        /**
+         * Returen the FEValues scalar extractor for this temperature
```
Comment (bangerth) [inline include/aspect/simulator.h]: -> Return

**Comment by bangerth** [comment]
Very nice. Are these all the places you could find that could benefit from this?

Out of curiosity, why does it depend on #1146? Aren't these two patches entirely unrelated?

**Comment by yinghe616** [comment]
Applied the new changes to other two places.

@bangerth ,  yes, #1145 does not depend on #1146 :p.  I first started at 1145, and was planning to implement 1146 until it's merged.  But now anyhow I need to rebase one of them later, right?:)

**Comment by yinghe616** [comment]
@bangerth , there is a function in melt.cc 
 353       const unsigned int solution_component
 354         = (advection_field.is_temperature()
 355            ?
 356            introspection.component_indices.temperature
 357            :
 358            introspection.component_indices.compositional_fields[advection_field.compositional_variable]
 359           );

which can be simplified as 
 const unsigned int solution_component = advection_field.component_index(introspection);

Do you suggest to update it in this pull request, or a separate one?

**Comment by bangerth** [comment]
Thanks, @yinghe616. I think you already included the place in `melt.cc` in your patch, right?

**Comment by yinghe616** [comment]
@bangerth , the other place I mentioned above is related to solution_component not the  FEValue scalar_extractor, so I didn't include it. But I can include it in the pull #1146

**Comment by bangerth** [comment]
If it's logically independent, put it into a separate (new) pull request. That typically makes it easiest to review things.

**Pull request merged**